### ISSUE TITLE
Add config option to disable WebSocket masking

### ISSUE THREAD

**Issue by crertel** [mtrudel/bandit#271]
This is a bit of a gnarly request, but I can justify it, sorta.

If you're already running SSL for your websockets, the whole masking part of the protocol is a lot less relevant I think.

Would it be possible to add a config mechanism to basically make the `mask/2` function in `Bandit.Websocket.Frame` ([here](https://github.com/mtrudel/bandit/blob/9b55f641facb2b9a719c49d79e28656dc6c2a441/lib/bandit/websocket/frame.ex#L187)) a no-op? The [RFC](https://www.rfc-editor.org/rfc/rfc6455#section-10.3) seems to be protecting against a threat model that doesn't really exist so much if SSL is guaranteed.

Obviously this would only work with clients that are expecting it, but for reducing overhead per-connection and frame it'd be nice--say, for a Raspberry Pi or embedded device.

**Comment by alisinabh** [comment]
Hey @crertel,

I would like to add a few notes here. (@mtrudel please feel free to correct this)

Firstly, when talking about reducing overhead we need to consider the impact that we are trying to minimize. I would love to see if you already have benchmarks for something like this. By just skimming though the masking code, I personally really doubt skipping it would bring a lot of improvements here (I might be wrong but again, we can benchmark).

Secondly, I don't think using a secure connection (wss) will eliminate this issue in all cases as the cache might live in your ingress layer and still prune to poisoning.

Finally and most importantly, masking is required part of RFC-6455§5.3 and if you remove it, well, it is no longer called a websocket. Even if you do remove it in bandit (or any web server), You also need to implement custom client code which does not mask the data sent to the server which is no longer RFC-6455 compatible.

**Comment by crertel** [comment]
Yep, I'm aware that this is departing from the spec--for the use-case I've got in mind, that's perfectly acceptable (and why this would be a totally optional thing).

The performance can be an issue on low-end embedded devices where touching all of the bytes and doing additional allocations for the mask is expensive. On desktop/server x86/x64/ARM, it's basically a non-issue.

You're correct that the resulting thing isn't an RFC-compliant websockets implementation and that it would be a weird thing for a niche usecase--that deviation from the standard is a feature, not a bug.

**Comment by mtrudel** [comment]
Back when I was doing Bandit's last perf workup (the 0.6.x series) I looked at this in some detail. Masking and UTF-8 validation were by far the two biggest contributors to WebSocket runtime, and both are unavoidable in practice (recall that Bandit's [literal prime directive is correctness](https://github.com/mtrudel/bandit/tree/main#readme)). @moogle19 and I ended up getting some decent wins [here](https://github.com/mtrudel/bandit/commit/6ab40a1ca44e7877b1c57d387727574d9c1e4957) and [here](https://github.com/mtrudel/bandit/pull/32), among others. All to say that the perf cost of masking is something I'm well aware of. Even to the point of making UTF-8 validation [optional](https://github.com/mtrudel/bandit/commit/88d505660ab033bc273e6035317dfb7ec86406e1). 

That being said, I don't think your ask is something I want to bring into the library. To me, this is different than UTF-8 validation because it fundamentally changes the wire protocol to not be RFC compliant (in contrast to UTF-8 validation which just removed a validation but still presumed a compliant client). This may be for valid reasons as you point out, but they really aren't ones that are universally applicable.

I see a couple of paths forward though:

1. @moogle19's early work on bandit-native support of NIF'ing hot paths would presumably tackle masking right off of the gate. This would necessitate a way to DI a masking implementation, meaning you could do this yourself nearly trivially.
2. If there were a matching client & compelling use case layered on top I'd be more inclined to entertain this (mind, with some sort of negotiation on top (note that using extensions for this would REQUIRE a registered extension name per [RFC6455§9.1](https://datatracker.ietf.org/doc/html/rfc6455#section-9.1)). 

I hope this is a worthy explanation of my rationale. I'm going close this issue for hygiene, but in the meantime if you're looking to push this forward I'd say that checking in with @moogle19 to see what plans he has for bandit-native is a logical next step.

Thanks for the issue!
